datafusion/functions/src/datetime/to_local_time.rs

```diff
@@ -0,0 +1,601 @@
+                        Some(adjusted_ts),
+                        None,
+                    )))
+                }
+                ColumnarValue::Scalar(ScalarValue::TimestampSecond(
+                    Some(ts),
+                    Some(tz),
+                )) => {
+                    let adjusted_ts =
+                        adjust_to_local_time::<TimestampSecondType>(ts, &tz);
+                    Ok(ColumnarValue::Scalar(ScalarValue::TimestampSecond(
+                        Some(adjusted_ts),
+                        None,
+                    )))
+                }
+                ColumnarValue::Array(array) => {
+                    fn transform_array<T>(
+                        array: &ArrayRef,
+                        tz: &str,
+                    ) -> Result<ColumnarValue>
+                    where
+                        T: ArrowTimestampType,
+                    {
+                        let array = as_primitive_array::<T>(array)?;
+                        let array: PrimitiveArray<T> =
+                            array.unary(|ts| adjust_to_local_time::<T>(ts, tz));
+
+                        Ok(ColumnarValue::Array(Arc::new(array)))
+                    }
+
+                    match array.data_type() {
+                        Timestamp(_, None) => {
+                            // if no timezone specificed, just return the input
+                            Ok(ColumnarValue::Array(Arc::new(array)))
+                        }
+                        Timestamp(Nanosecond, Some(tz)) => {
+                            transform_array::<TimestampNanosecondType>(&array, tz)
+                        }
+                        Timestamp(Microsecond, Some(tz)) => {
+                            transform_array::<TimestampMicrosecondType>(&array, tz)
+                        }
+                        Timestamp(Millisecond, Some(tz)) => {
+                            transform_array::<TimestampMillisecondType>(&array, tz)
+                        }
+                        Timestamp(Second, Some(tz)) => {
+                            transform_array::<TimestampSecondType>(&array, tz)
+                        }
+                        _ => {
+                            exec_err!("to_local_time function requires timestamp argument in array, got {:?}", array.data_type())
+                        }
+                    }
+                }
+                _ => {
+                    exec_err!(
+                        "to_local_time function requires timestamp argument, got {:?}",
+                        time_value.data_type()
+                    )
+                }
+            },
+            _ => {
+                exec_err!(
+                    "to_local_time function requires timestamp argument, got {:?}",
+                    arg_type
+                )
+            }
+        }
+    }
+}
+
+/// This function converts a timestamp with a timezone to a timestamp without a timezone.
```

Comment:
This is one of the clearest explanations of what a function does that I have read in a long time 💯  Nice work @appletreeisyellow